src/image/TransmissionCurve.cc

```diff
@@ -0,0 +1,851 @@
+        auto wlIter = wavelengths.begin();
+        for (auto outIter = out.begin(); outIter != out.end(); ++outIter, ++wlIter) {
+            double & y = *outIter;
+            if (*wlIter < _wavelengths.front()) {
+                y = _atBounds.first;
+            } else if (*wlIter > _wavelengths.back()) {
+                y = _atBounds.second;
+            } else {
+                int status = ::gsl_interp_eval_e(_interp.get(), _wavelengths.getData(), _throughput.getData(),
+                                                 *wlIter, accel.get(), &y);
+                if (status) {
+                    throw LSST_EXCEPT(
+                        pex::exceptions::InvalidParameterError,
+                        (boost::format("Invalid wavelength value: %g") % (*wlIter)).str()
+                    );
+                }
+            }
+        }
+    }
+
+    bool isPersistable() const override { return true; }
+
+protected:
+
+    // transforming a ConstantTransmissionCurve is a no-op
+    std::shared_ptr<TransmissionCurve> _transformImpl(
+        std::shared_ptr<TransmissionCurve> self,
+        std::shared_ptr<geom::TransformPoint2ToPoint2> transform
+    ) const override {
+        return self;
+    }
+
+    std::string getPersistenceName() const override { return "ConstantTransmissionCurve"; }
+
+    std::string getPythonModule() const override { return "lsst.afw.image"; }
+
+    void write(OutputArchiveHandle& handle) const override;
+
+private:
+    std::pair<double,double> _atBounds;
+    ndarray::Array<double,1,1> _throughput;
+    ndarray::Array<double,1,1> _wavelengths;
+    int _naturalSamplingSize;
+    CPtr<::gsl_interp> _interp;
+};
+
+struct ConstantPersistenceHelper {
+    table::Schema schema;
+    table::Key<double> throughputAtMin;
+    table::Key<double> throughputAtMax;
+    table::Key<table::Array<double>> throughput;
+    table::Key<table::Array<double>> wavelengths;
+
+    static ConstantPersistenceHelper const & get() {
+        static ConstantPersistenceHelper const instance;
+        return instance;
+    }
+
+private:
+
+    ConstantPersistenceHelper() :
+        schema(),
+        throughputAtMin(schema.addField<double>(
+            "throughputAtMin", "throughput below minimum wavelength")),
+        throughputAtMax(schema.addField<double>(
+            "throughputAtMax", "throughput above minimum wavelength")),
+        throughput(schema.addField<table::Array<double>>(
+            "throughput", "array of known throughput values", "", 0)),
+        wavelengths(schema.addField<table::Array<double>>(
+            "wavelengths", "array of wavelengths corresponding to known throughputs", "angstrom", 0))
+    {
+        schema.getCitizen().markPersistent();
+    }
+
+};
+
+void ConstantTransmissionCurve::write(OutputArchiveHandle & handle) const {
+    auto const & keys = ConstantPersistenceHelper::get();
+    table::BaseCatalog catalog = handle.makeCatalog(keys.schema);
```

Comment:
`makeCatalog` may modify `handle`; what happens if a later call (e.g., to `set`) throws an exception?

Reply:
Archive may be corrupted; I think that's consistent with your recommendation to mark `OutputArchive.put` (which delegates to `write` methods like this one) as providing no exception safety.  Did you have something in mind beyond that?